### PR TITLE
Implement `AsyncAurora` contract

### PR DESCRIPTION
**Pull request by karim-en**
Implement `AsyncAurora` contract to support [Aurora Cross Contract Calls](https://github.com/aurora-is-near/aurora-engine/discussions/461)

### CODE REVIEW

**Review by sept-en — CHANGES_REQUESTED** [review]
Some minor comments while it's still draft. More comments will be added when it's not a draft anymore

**Comment by joshuajbouw** [comment]
Just be aware, that we need to have DAO approval for this.

File: etc/async-aurora/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                        promise = promises[combinator.promise_index as usize]
+                            .clone()
+                            .then(promise);
+                    }
+                }
+            }
+
+            promises.push(promise);
+        }
+    }
```
Comment (sept-en) [inline etc/async-aurora/src/lib.rs:97]: For a discussion: do we want to have lazy-promise execution? This way we may need to ask the user to attach enough ETH to pay for the storage needed to store `promises` in `async.aurora` before they are executed. Also, we will need to have a precompile allowing the user to execute unfinished promises that are already stored in `async.aurora` in some mapping (`execution_id: Vec<Promise>`)

File: etc/async-aurora/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
+mod engine_types;
+use crate::engine_types::*;
+mod async_promise;
+use crate::async_promise::*;
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::{assert_self, env, ext_contract, near_bindgen, Gas, Promise};
+
+pub const GAS_RESERVED_FOR_CURRENT_CALL: Gas = 20_000_000_000_000;
```
Comment (sept-en) [inline etc/async-aurora/src/lib.rs:8]: Suggesting changing the name. Also, are you sure 20 TGas will be enough? What's the real usage value?
```suggestion
pub const GAS_RESERVED_FOR_SUBMIT: Gas = 20_000_000_000_000;
```

File: etc/async-aurora/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "async-aurora"
+version = "0.1.0"
+authors = ["Aurora <hello@aurora.dev>"]
+edition = "2021"
+publish = false
+
```
Comment (sept-en) [inline etc/async-aurora/Cargo.toml:7]: let's also add the description.

**Review by sept-en — CHANGES_REQUESTED** [review]
Added some more comments and questions for discussions. Also, please add README for async-aurora with appropriate references to cross-contract-call and native token connector design discussions.

File: engine-precompiles/src/async_router.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+///
+/// Address: `0xad65a767211ae644cdf2d036853e2bcda225dff8`
+/// This address is computed as: `&keccak("asyncRouter")[12..]`
+pub const ADDRESS: Address = crate::make_address(0xad65a767, 0x211ae644cdf2d036853e2bcda225dff8);
+
+fn predecessor_address(predecessor_account_id: &AccountId) -> Address {
+    near_account_to_evm_address(predecessor_account_id.as_bytes())
+}
+
+impl Precompile for AsyncRouter {
```
Comment (sept-en) [inline engine-precompiles/src/async_router.rs:34]: Another idea would be to implement `AsyncRouter` as a regular Solidity contract and create a precompile that will call that router contract. The router contract could be deployed using a separate function (e.g. `deploy_router_contract()`.

File: etc/async-aurora/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        let mut promises: Vec<Promise> = Vec::new();
+        for (ix, promise_desc) in promises_desc.iter().enumerate() {
+            let promise = Promise::new(promise_desc.target.clone()).function_call(
+                promise_desc.method_name.clone(),
+                promise_desc.arguments.clone(),
+                0,
+                promise_desc.gas,
+            );
+
+            promises.push(promise.clone());
```
[the remark this answers is not in the document]
Comment (karim-en) [reply etc/async-aurora/src/lib.rs]: Fixed at 77836d8420f8b720f7c2718373528ae05f68136e

File: etc/async-aurora/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+#[ext_contract(ext_bridge_token_factory)]
+pub trait ExtBridgeTokenFactory {
+    #[result_serializer(borsh)]
+    fn finish_withdraw(
+        &self,
+        #[serializer(borsh)] amount: Balance,
+        #[serializer(borsh)] recipient: AccountId,
+    ) -> Promise;
+}
+
```
[the remark this answers is not in the document]
Comment (karim-en) [reply etc/async-aurora/src/lib.rs]: Fixed at acbf0b07d46e663159c583e1a569d50eba0c6ef9

File: engine-precompiles/src/async_router.rs (hunk 4 above)
Comment (karim-en) [reply engine-precompiles/src/async_router.rs:34]: Yes, this is the point of contention, I also preferred the regular solidity contract, but I got request to implement it as precompile

File: etc/async-aurora/src/lib.rs (hunk 2 above)
Comment (karim-en) [reply etc/async-aurora/src/lib.rs]: Fixed at bfe58440b083f686a714d95d5c4a968967c2bcdc

File: engine-tests/src/tests/res/TestAsync.sol [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+pragma solidity ^0.8;
+
+contract TestAsync {
+    constructor() public {}
+
+    function simpleCall(
+        string memory accountId,
+        string memory method,
+        uint128 arg,
+        string memory gas
```
[the remark this answers is not in the document]
Comment (karim-en) [reply engine-tests/src/tests/res/TestAsync.sol]: Fixed at d4e17fcf6164f59a05c870c2b3a6e1a96d7eafbd

File: engine-tests/src/tests/async_aurora.rs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        TestContext {
+            async_aurora,
+            caller,
+            receiver,
+            aurora,
+            chain_id,
+        }
+    }
+
+    fn submit(context: &TestContext, input: Vec<u8>) {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply engine-tests/src/tests/async_aurora.rs]: fixed at bfe58440b083f686a714d95d5c4a968967c2bcdc

File: etc/async-aurora/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+struct PromiseDescription {
+    target: AccountId,
+    method_name: Vec<u8>,
+    arguments: Vec<u8>,
+    gas: u64,
+    combinator: Option<CombinatorDescription>,
+}
+
+fn parse_promises(input: String) -> Vec<PromiseDescription> {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply etc/async-aurora/src/lib.rs]: fixed at acbf0b07d46e663159c583e1a569d50eba0c6ef9

File: engine-tests/src/tests/async_aurora.rs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+                context.async_aurora.account_id(),
+                "submit",
+                &submit_args.try_to_vec().unwrap(),
+                near_sdk_sim::DEFAULT_GAS,
+                0,
+            )
+            .assert_success();
+    }
+
+    fn caller_simple_call(context: &TestContext, method: String, arg: u128) {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply engine-tests/src/tests/async_aurora.rs]: Fixed at f004ca6d67fd0724e03495fb0386ebac11132ce5

File: etc/async-aurora/Cargo.toml (hunk 3 above)
Comment (karim-en) [reply etc/async-aurora/Cargo.toml]: Fixed at f004ca6d67fd0724e03495fb0386ebac11132ce5

File: etc/async-aurora-test/src/engine_types.rs [hunk 11]
```diff
@@ -0,0 +1,53 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+
+pub type EthAddress = [u8; 20];
+pub type RawU256 = [u8; 32];
+/// Wei compatible Borsh-encoded raw value to attach an ETH balance to the transaction
+pub type WeiU256 = [u8; 32];
+
```
Comment (sept-en) [inline etc/async-aurora-test/src/engine_types.rs:7]: NB: this file is being replicated twice in both `etc/async-aurora-test` and `etc/async-aurora`. Moreover, it shouldn't be duplicated there, rather should be separated into a separate lib in the repo itself.

**Comment by bharath-123** [comment]
Would this also cover querying a NEAR contract? It is great that we can call a NEAR contract, but i would like to understand if this feature would allow us to query a NEAR contract method too?

**Comment by birchmd** [comment]
Closing this as cross-contract calls were implemented in #560

**Pull request closed without merging**